errors, repl: migrate to use internal/errors.js

no23reason opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

Migrated `repl` and `internal/repl` to `internal/errors`. Uses some code from #11294 as recommended in #11273.

@jasnell, pinging you for mentoring as this is my first PR to this project, thank you!

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
errors, REPL

joyeecheung commented on doc/api/errors.md:
```diff
@@ -563,6 +570,67 @@ found [here][online].
+<a id="ERR_REPL_OLD_HISTORY_PARSING_ERROR"></a>
+### ERR_REPL_OLD_HISTORY_PARSING_ERROR
+
+The `'ERR_REPL_OLD_HISTORY_PARSING_ERROR'` error code is used when parsing
```
Probably `ERR_INVALID_OLD_REPL_HISTORY`? The `_ERROR` at the end is unnecessary.

no23reason replied:
You are right, fixed it. Thanks!

joyeecheung commented on doc/api/errors.md:
```diff
@@ -563,6 +570,67 @@ found [here][online].
+<a id="ERR_REPL_INVALID_EVAL_CONFIG"></a>
+### ERR_REPL_INVALID_EVAL_CONFIG
+
+The `'ERR_REPL_INVALID_EVAL_CONFIG'` error code is used when both
```
The error being thrown in the code is in fact `ERR_INVALID_REPL_EVAL_CONFIG`(`REPL` and `INVALID` are switched). 

Also I think `s/config/options` because that's what is written down in the docs, and those are in fact optional? `ERR_INVALID_REPL_EVAL_OPTION` with the current error message should be better.

no23reason replied:
The rationale behind the two codes was to group the REPL-related errors together. However your points are very valid, so I fixed it according to your suggestions. Thank you very much for your feedback!

fhinkel commented on doc/api/errors.md:
```diff
@@ -563,6 +570,67 @@ found [here][online].
+
+The `'ERR_INVALID_REPL_EVAL_OPTION'` error code is used when both
+`breakEvalOnSigint` and `eval` options are set in the REPL config.
+Allowing both of these would be confusing as `breakEvalOnSigint`
```
Is `confusing` the right word here? 

How about something like this?
`The `'ERR_INVALID_REPL_EVAL_OPTION'` error code is used when both
 `breakEvalOnSigint` and `eval` options are set in the REPL config, which is not supported.

no23reason replied:
I was trying to explain the rationale of it not being supported, but you are right, it is not the place to do it. I changed the formulation.

fhinkel commented:
@no23reason Thanks so much for putting this together. Sorry that it is dragging out for so long due to being a semver-major change. Could you rebase and also squash your commits (I think all the changes should be one commit, right?). Thanks!

no23reason commented:
No problem, I'll get to it tomorrow afternoon. Thank you for your responses!

On 23 May 2017, at 18:49, F. Hinkelmann <notifications@github.com<mailto:notifications@github.com>> wrote:


@no23reason<https://github.com/no23reason> Thanks so much for putting this together. Sorry that it is dragging out for so long due to being a semver-major change. Could you rebase and also squash your commits (I think all the changes should be one commit, right?). Thanks!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/nodejs/node/pull/11347#issuecomment-303463311>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ABqurUgPtDdBINGoFT0HhO-wmpoAP74Qks5r8w4lgaJpZM4L_ich>.

no23reason commented:
@fhinkel I rebased, squashed and implemented the change you proposed.
However the tests are now failing on files I believe have nothing to do with my changes:

> test\parallel\test-cluster-inspector-debug-port.js
> test\sequential\test-net-connect-local-error.js
> test\sequential\test-net-server-address.js

I'm not sure how to proceed now, can you please help me?

joyeecheung commented:
@no23reason 
`test-cluster-inspector-debug-port.js` is in https://github.com/nodejs/node/issues/12376 so that's probably the reason of failure. The rest I am not quite sure especially the flakiness of `test-net-connect-local-error` should have been addressed..

Fired up a CI to see how it goes anyway:
https://ci.nodejs.org/job/node-test-pull-request/8295/

refack commented:
> test\sequential\test-net-connect-local-error.js

@no23reason can you help me with the error you're getting for this test?
1. Paste the error output of the test (you can just run `Release\node.exe test\sequential\test-net-connect-local-error.js`, if it doesn't reproduce run `python tools\test.py -v sequential` to run the whole suite since sometimes tests interact with each other)
2. Make sure commit 3429c90 is in your tree

no23reason commented:
OK, so today I tried the build again on a clean checkout of the repo and suddenly all the tests pass except the `test\parallel\test-cluster-inspector-debug-port.js`
So I guess everything is OK now, right? Sorry for the false alarm

refack commented:
> Sorry for the false alarm

NP. Dirty builds happen all the time 🤷‍♂️ ... Especially on Windows

joyeecheung commented on doc/api/errors.md:
```diff
@@ -801,8 +814,9 @@ likely an indication of a bug within Node.js itself.
 [syscall]: http://man7.org/linux/man-pages/man2/syscall.2.html
 [try-catch]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch
-[vm]: vm.html
+[vm]: vm.html
```
Accidental change?

no23reason replied:
Yes, my editor trolled me a bit, I fixed it, thank you!

joyeecheung commented on doc/api/errors.md:
```diff
@@ -801,8 +814,9 @@ likely an indication of a bug within Node.js itself.
 [domains]: domain.html
 [event emitter-based]: events.html#events_class_eventemitter
 [file descriptors]: https://en.wikipedia.org/wiki/File_descriptor
+[Node.js Error Codes]: #nodejs-error-codes
```
This is a duplicate of an existing ref..

no23reason replied:
Thanks, good catch, I removed it

jasnell commented on lib/internal/errors.js:
```diff
@@ -136,6 +138,8 @@ E('ERR_IPC_DISCONNECTED', 'IPC channel is already disconnected');
 E('ERR_IPC_SYNC_FORK', 'IPC cannot be used with synchronous forks');
 E('ERR_MISSING_ARGS', missingArgs);
+E('ERR_INVALID_REPL_EVAL_OPTION',
+  'Cannot specify both breakEvalOnSigint and eval for REPL');
```
Can you put the new codes in as close to alphabetic order as possible :-)

no23reason replied:
Of course, sorry I missed it :)

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/8369/

joyeecheung commented:
https://github.com/nodejs/node/pull/13299 has already landed so unfortunately we can not get this PR in anymore . Thanks for the contribution anyway!

joyeecheung commented:
Oh, looks like #13299 only contains part of what this PR does, so this still adds something that isn't in the codebase yet. @no23reason can do a rebase again? Thanks!

no23reason commented:
@joyeecheung I rebased and removed the error code that is now obsolete.

tniessen commented on lib/internal/errors.js:
```diff
@@ -129,6 +129,8 @@ E('ERR_INVALID_OPT_VALUE',
     return `The value "${String(value)}" is invalid for option "${name}"`;
   });
+E('ERR_INVALID_REPL_EVAL_CONFIG',
+  'Cannot specify both breakEvalOnSigint and eval for REPL');
```
I think we agreed to use quotation marks around parameter names in error message (`'Cannot specify both "breakEvalOnSigint" and "eval" for REPL'`), see #1220 (or literally three lines above in `errors.js`). I am not 100% sure whether this applies here, can someone confirm?

Anyway, that would make this PR semver-major, so feel free to ignore it in order to get your PR in soon :)

no23reason replied:
I was using the original text of the error, but I agree that it should use the quotation marks (I wasn't aware of the convention). Thanks!

tniessen replied:
Thank you for fixing it quickly! As I said before, this will (sadly) make this PR semver-major, meaning that we probably won't be able to release it before node 9.

no23reason replied:
It's ok :) It was labelled as such even before the 8.0 came out so I guess it kind of stuck there...

no23reason commented on doc/api/errors.md:
```diff
@@ -624,6 +624,13 @@ process. See [`child.send()`] and [`process.send()`] for more information.
+<a id="ERR_INVALID_REPL_EVAL_CONFIG"></a>
+### ERR_INVALID_REPL_EVAL_CONFIG
+
+The `'ERR_INVALID_REPL_EVAL_CONFIG'` error code is used when both
```
I changed it, thank you for pointing me to the PR :)

tniessen commented:
CI before merging: https://ci.nodejs.org/job/node-test-pull-request/8661/

tniessen commented:
Landed in 3ee37329da7323a86c248a63fe20d40885c7a84a :tada: Thank you for your first contribution!

refack commented:
For my bookkeeping: #13686

tniessen commented:
Landed in aff8d358fabfc6bd30c48c1756453e49ad984b5f, thank you for your first contribution! :tada: We'd be happy to see more!

(And sorry for closing this prematurely, mixed this up with another PR.)

no23reason commented:
@tniessen Thank you for your patience and advice, I hope my next contributions will be smoother :)
